File: frontend/src/models/api/Address.ts
import { Api_ConcurrentVersion } from './ConcurrentVersion';
import { Api_Organization } from './Organization';
import { Api_Person } from './Person';
import Api_TypeCode from './TypeCode';

export interface Api_Country extends Api_ConcurrentVersion {
  countryId?: number;
  countryCode?: string;
  description?: string;
}

export interface Api_ProvinceState extends Api_ConcurrentVersion {
  provinceStateId?: number;
  provinceStateCode?: string;
  description?: string;
}

export interface Api_Address extends Api_ConcurrentVersion {
  id?: number;
  streetAddress1?: string;
  streetAddress2?: string;
  streetAddress3?: string;
  municipality?: string;
  province: Api_ProvinceState;
  country?: Api_Country;
  countryOther?: string | null;
  postal?: string;
}

export interface Api_OrganizationAddress extends Api_ConcurrentVersion {
  id?: number;
  isDisabled?: boolean;
  organization?: Api_Organization;
  address?: Api_Address;
  addressUsageType?: Api_TypeCode<string>;
}

export interface Api_PersonAddress extends Api_ConcurrentVersion {
  id?: number;
  isDisabled?: boolean;
  person?: Api_Person;
  address?: Api_Address;
  addressUsageType?: Api_TypeCode<string>;
}

File: source/backend/entities/Partials/Address.cs
using System;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;

namespace Pims.Dal.Entities
{
    /// <summary>
    /// Address class, provides an entity for the datamodel to manage property addresses.
    /// </summary>
    public partial class PimsAddress : StandardIdentityBaseAppEntity<long>, IBaseAppEntity
    {
        #region Properties
        [NotMapped]
        public override long Internal_Id { get => this.AddressId; set => this.AddressId = value; }
        #endregion

        #region Constructors

        /// <summary>
        /// Create a new instance of a Address class.
        /// </summary>
        /// <param name="address1"></param>
        /// <param name="municipality"></param>
        /// <param name="province"></param>
        /// <param name="district"></param>
        /// <param name="postal"></param>
        public PimsAddress(string address1, string address2, string municipality, PimsProvinceState province, PimsDistrict district, string postal)
            : this()
        {
            if (string.IsNullOrWhiteSpace(address1))
            {
                throw new ArgumentException($"Argument '{nameof(address1)}' is required.", nameof(address1));
            }

            this.StreetAddress1 = address1;
            this.StreetAddress2 = address2;
            this.MunicipalityName = municipality ?? throw new ArgumentNullException(nameof(municipality));
            this.ProvinceState = province ?? throw new ArgumentNullException(nameof(province));
            this.ProvinceStateId = province.ProvinceStateId;
            this.Country = province.Country ?? throw new ArgumentException($"Argument '{nameof(province)}.{nameof(province.Country)}' is required.", nameof(province));
            this.CountryId = province.Country.CountryId;
            this.DistrictCodeNavigation = district;
            this.DistrictCode = district?.DistrictCode;
            this.RegionCodeNavigation = district?.RegionCodeNavigation;
            this.RegionCode = district?.RegionCode;
            this.PostalCode = postal;
        }
        #endregion

        #region Methods

        /// <summary>
        /// Return the address as a string.
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            return string.Join(", ", new[] { this.StreetAddress1, this.StreetAddress2, this.StreetAddress3, this.MunicipalityName, this.ProvinceState?.ProvinceStateCode, this.DistrictCodeNavigation?.DistrictName, this.RegionCodeNavigation?.RegionName, this.PostalCode, this.Country?.CountryCode }.Where(s => !string.IsNullOrWhiteSpace(s)));
        }
        #endregion
    }
}
